ghcide/src/Development/IDE/Plugin/Completions.hs

```diff
@@ -152,13 +156,28 @@ getCompletionsLSP ide plId
               (Just pfix', _) -> do
                 let clientCaps = clientCapabilities $ shakeExtras ide
                 config <- getCompletionsConfig plId
-                allCompletions <- liftIO $ getCompletions plId ideOpts cci' parsedMod bindMap pfix' clientCaps config
+                allCompletions <- liftIO $ getCompletions plId ideOpts cci' parsedMod bindMap pfix' clientCaps config moduleExports
                 pure $ InL (List allCompletions)
               _ -> return (InL $ List [])
           _ -> return (InL $ List [])
       _ -> return (InL $ List [])
 
 ----------------------------------------------------------------------------------------------------
+  
+buildLocalModuleExports:: IdeState -> ([(Located ModuleName, Maybe ArtifactsLocation)], PositionMapping) -> IO (Map.HashMap T.Text (Set.HashSet IdentInfo))
+buildLocalModuleExports ide inMap = do
+  mModIfaces <- mapM (getModIface ide) $ fst inMap
+  pure (buildModuleExportMapFrom $ catMaybes mModIfaces)
+
+getModIface :: IdeState ->  (Located ModuleName, Maybe ArtifactsLocation) -> IO (Maybe ModIface)
+getModIface ide item = do
+  let nPath = artifactFilePath <$> snd item
+  case nPath of 
+    (Just path) -> do 
+      file <- runIdeAction "Completion" (shakeExtras ide) $ do
+        fmap fst <$> useWithStaleFast GetModIface path
+      pure $ fmap hirModIface file
+    Nothing -> pure Nothing
```

Comment:
For performance reasons, can you restructure this code so that it performs only one `runIdeAction` and one call to `usesWithStaleFast`?

Reply:
Thank you for all of your help on this PR! Your pointers have been extremely helpful.

I have a question here - Do I not need to call these for every path returned from  `GetLocatedImports`?

Reply:
If you knew which import is being autocompleted then you could query only that one.

Reply:
I was thinking as a followup to this PR, I could improve the completion of imports to also include local modules. In order to do this, I can use the keys from `moduleExportsMap` in `Logic.hs` rather than `importableModules`. Do you think this is a good direction? And if so, does it make sense to query for all located imports so we have this feature?

Reply:
> I was thinking as a followup to this PR, I could improve the completion of imports to also include local modules. In order to do this, I can use the keys from `moduleExportsMap` in `Logic.hs` rather than `importableModules`. Do you think this is a good direction? 

Or you could just use the list of imports, which can be extracted from the parsed module. 

> And if so, does it make sense to query for all `GetLocatedImports` so we have this feature?

Querying GetModIface for all located imports and then building a Map is O(nlogn) for the number of imports, but I'm not too worried since in most cases n will be small. So it does make sense, it's just a bit inefficient.

Reply:
But please, make sure to query all the GetModIface keys in a single call of usesWithStale so that they all happen in parallel

Reply:
addressed